python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+
+The high-level goals for the Python SDK are:
+
+* Be as simple as possible from a caller POV
+* Be as lightweight as possible wrt dependencies
+* Be reasonably flexible with reasonable defaults
+* Support only Python >= 3.7
+* Be friendly to REPL/interpreter use cases
+
+The stack/approach looks like:
+
+* Python >= 3.7 (so may need `typing_extension` for things like `ParamSpec`)
+* Poetry packaging, black formatting
+  * Pipenv originally considered
+* PyO3 + sdk-core
+* asyncio direct (i.e. no need for anyio-style wrapper)
+* Type annotations everywhere
+* Decorator-based activity and workflow definitions
+
+An example of a client (intentionally missing _a lot_ of features):
+
+```python
+import temporalio
+
+def main():
+  client = temporalio.Client("localhost:7233", namespace="MyNamespace")
+  result = await client.execute_workflow("MyWorkflow", "arg1", id="someid", task_queue="somequeue")
+  print("Result", result)
+```
+
+An example of an activity and registration (also missing a lot of features):
+
+```python
+import temporalio
+
+def say_hello(name: str) -> str:
+  return f"Hello, {name}!"
+
+def main(client: temporalio.Client):
+  worker = temporalio.Worker(client, task_queue="somequeue")
+  worker.register_activity(say_hello)
+  await worker.run()
+```
+
+## API Definitions
+
+A lot of loosely-defined API definitions are below. They are intentionally lacking details.
+
+High-level notes:
+
+* For demonstration/simplicity here we are leaving off a lot of proper typings (e.g. `Optional`) that the real version
+  will have
+* We leave off a lot of options on calls for this high-level proposal
+* We are not using proper function documentation, but rather inline comments for easier discussion
+* **NOTE: These will have proper docs and types when developed!**
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
+exported from `temporalio`):
+
+```python
+class Client:
+    def __init__(
+        self,
+        addr: str,
+        *,
+        namespace: str = "default",
+        default_task_queue: str = None,
+        identity: str = None
+        # ... additional options omitted for brevity
+    ) -> None:
+        # Actually dials gRPC here
+        pass
+
+    # Raw gRPC service
+    @property
+    def service(self) -> workflowservice.WorkflowService:
+        pass
+
+    def start_workflow(
+        workflow: str,
+        *args,
+        id: str = None,
+        task_queue: str = None,
+        retry_policy: RetryPolicy = None,
+        # Note, we are embedding signal-with-start in here
+        start_signal: str = None,
+        start_signal_args=[],
+        # ... additional options omitted for brevity
+    ) -> WorkflowHandle[Any]:
+        pass
+
+    # In a future phase:
+    # @overload
+    # def start_workflow(workflow: Callable[P, T], *args: P.args, ...) -> WorkflowExecution[T]
+
+    # Does not validate whether a workflow exists
+    def workflow(
```

Comment:
1. I'd rather stick to the name `get_handle`
2. This should become generic too (if possible)

Reply:
> I'd rather stick to the name get_handle

Can you provide justification? Running `await client.workflow("myworkflow")` is very clear software. I am really trying to make the code look great from a caller perspective but I fear that we're more concerned with other things.

> This should become generic too (if possible)

It would definitely be generic later. I provided an overload for the previous one just for thought, I do not provide all possible overloads throughout this review.

Reply:
`workflow` is confusing, it's not a verb so it's unclear what it actually does.
Additionally this method should not be async since it does not do any IO.

I'm fine with not providing all of the overloads, didn't mean for this to be addressed in this PR.
We'll revisit later when we get to defining the workflow API

Reply:
It's half getter half creator (you are creating the handle, not getting it). Basically, you might as well construct the workflow handler via constructor w/ a client and workflow ID but obviously that's uglier. Often when it's just a simple constructor, one expects to leave the `new_` or `create_` or whatever. And in order to disambiguate, I wanted the word `workflow` in there.

How about `get_workflow`? Or would you rather `get_workflow_handle`?

Reply:
I have changed this to `get_workflow_handle`